Operator panics when changing from ephemeral storage to PVCs

Issue was reported in Opni - https://github.com/rancher/opni/issues/1700
```[04:08:04] INFO Generating certificates {"controller": "opensearchcluster", "controllerGroup": "[opensearch.opster.io](http://opensearch.opster.io/)", "controllerKind": "OpenSearchCluster", "OpenSearchCluster": {"name":"opni","namespace":"opni"}, "namespace": "opni", "name": "opni", "reconcileID": "11a160f3-966e-4979-b16f-b5848f1a51a6", "interface": "http"}
[04:08:04] INFO updating existing ism {"controller": "multiclusterrolebinding", "controllerGroup": "[logging.opni.io](http://logging.opni.io/)", "controllerKind": "MulticlusterRoleBinding", "MulticlusterRoleBinding": {"name":"opni","namespace":"opni"}, "namespace": "opni", "name": "opni", "reconcileID": "c3431a0e-56c2-4e53-b119-21012136273a", "policy": "log-policy"}
[04:08:04] INFO Observed a panic in reconciler: runtime error: index out of range [0] with length 0 {"controller": "opensearchcluster", "controllerGroup": "[opensearch.opster.io](http://opensearch.opster.io/)", "controllerKind": "OpenSearchCluster", "OpenSearchCluster": {"name":"opni","namespace":"opni"}, "namespace": "opni", "name": "opni", "reconcileID": "11a160f3-966e-4979-b16f-b5848f1a51a6"}
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0
goroutine 880 [running]:
[sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1())
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:115](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:115) +0x1e5
panic({0x488cea0?, 0xc003f9b260?})
        runtime/panic.go:914 +0x21f
[opensearch.opster.io/pkg/reconcilers.(*ClusterReconciler).reconcileNodeStatefulSet(_](http://opensearch.opster.io/pkg/reconcilers.(*ClusterReconciler).reconcileNodeStatefulSet(_), {{0xc001b3818c, 0x4}, 0x3, {0xc001b38190, 0x4}, {0xc0016ffc20, 0xc0016ffc50, {0x0, 0x0, ...}}, ...}, ...)
        opensearch.opster.io@v0.0.0-00010101000000-000000000000/pkg/reconcilers/cluster.go:308 +0x1f18
[opensearch.opster.io/pkg/reconcilers.(*ClusterReconciler).Reconcile(0xc00046abd0)](http://opensearch.opster.io/pkg/reconcilers.(*ClusterReconciler).Reconcile(0xc00046abd0))
        opensearch.opster.io@v0.0.0-00010101000000-000000000000/pkg/reconcilers/cluster.go:120 +0x9b0
[opensearch.opster.io/controllers.(*OpenSearchClusterReconciler).reconcilePhaseRunning(0xc001030a50](http://opensearch.opster.io/controllers.(*OpenSearchClusterReconciler).reconcilePhaseRunning(0xc001030a50), {0x63c5490, 0xc0016ff890})
        opensearch.opster.io@v0.0.0-00010101000000-000000000000/controllers/opensearchController.go:320 +0x7ad
[opensearch.opster.io/controllers.(*OpenSearchClusterReconciler).Reconcile(0xc001030a50](http://opensearch.opster.io/controllers.(*OpenSearchClusterReconciler).Reconcile(0xc001030a50), {0x63c5490, 0xc0016ff890}, {{{0xc001b380e4, 0x4}, {0xc001b380e0, 0x4}}})
        opensearch.opster.io@v0.0.0-00010101000000-000000000000/controllers/opensearchController.go:141 +0x779
[sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x63c5490](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x63c5490)?, {0x63c5490?, 0xc0016ff890?}, {{{0xc001b380e4?, 0x40b50a0?}, {0xc001b380e0?, 0x6399528?}}})
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:118](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:118) +0xb7
[sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000fafb80](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000fafb80), {0x63c54c8, 0xc000f65860}, {0x45f69c0?, 0xc001b8c020?})
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:314](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:314) +0x365
[sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000fafb80](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000fafb80), {0x63c54c8, 0xc000f65860})
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265) +0x1c9
[sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2())
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226) +0x79
created by [sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2) in goroutine 375
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:222](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:222) +0x565
```

The issue is occurring because the resize logic expects a PVC and a VolumeClaim template to already exist in the statefulset.  When this is the case we need to put in logic for updating the stateful set

--- a/opensearch-operator/pkg/helpers/helpers.go
+++ b/opensearch-operator/pkg/helpers/helpers.go
@@ -432,11 +432,11 @@ func WorkingPodForRollingRestart(ctx context.Context, k8sClient client.Client, s
 	replicas := lo.FromPtrOr(sts.Spec.Replicas, 1)
 	// Handle the simple case
 	if replicas == sts.Status.UpdatedReplicas+sts.Status.CurrentReplicas {
-		ordinal := replicas - 1 - sts.Status.UpdatedReplicas
+		ordinal := sts.Status.UpdatedReplicas
 		return ReplicaHostName(*sts, ordinal), nil
 	}
 	// If there are potentially mixed revisions we need to check each pod
-	for i := replicas - 1; i >= 0; i-- {
+	for i := int32(0); i < replicas; i++ {
 		podName := ReplicaHostName(*sts, i)
 		pod := &corev1.Pod{}
 		if err := k8sClient.Get(ctx, types.NamespacedName{Name: podName, Namespace: sts.Namespace}, pod); err != nil {

--- a/opensearch-operator/pkg/reconcilers/cluster.go
+++ b/opensearch-operator/pkg/reconcilers/cluster.go
@@ -13,6 +13,7 @@ import (
 	"github.com/cisco-open/k8s-objectmatcher/patch"
 	"github.com/cisco-open/operator-tools/pkg/reconciler"
 	"github.com/go-logr/logr"
+	"github.com/samber/lo"
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
@@ -311,55 +312,9 @@ func (r *ClusterReconciler) reconcileNodeStatefulSet(nodePool opsterv1.NodePool,
 	//Default is PVC, or explicit check for PersistenceSource as PVC
 	// Handle volume resizing, but only if we are using PVCs
 	if nodePool.Persistence == nil || nodePool.Persistence.PersistenceSource.PVC != nil {
-		if nodePool.DiskSize == "" { // Default case
-			nodePool.DiskSize = builders.DefaultDiskSize
-		}
-
-		existingDisk := *existing.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage()
-		nodePoolDiskSize, err := resource.ParseQuantity(nodePool.DiskSize)
+		err := r.maybeUpdateVolumes(existing, nodePool)
 		if err != nil {
-			r.logger.Error(err, fmt.Sprintf("Invalid diskSize '%s' for nodepool %s", nodePool.DiskSize, nodePool.Component))
 			return result, err
-		}
-
-		if existingDisk.Equal(nodePoolDiskSize) {
-			r.logger.Info("The existing disk size " + existingDisk.String() + " is same as passed in disk size " + nodePoolDiskSize.String())
-		} else {
-			r.logger.Info("Disk sizes differ for nodePool %s: current: %s, desired: %s", nodePool.Component, existingDisk.String(), nodePoolDiskSize.String())
-			annotations := map[string]string{"cluster-name": r.instance.GetName()}
-			r.recorder.AnnotatedEventf(r.instance, annotations, "Normal", "PVC", "Starting to resize PVC %s/%s from %s to  %s ", existing.Namespace, existing.Name, existingDisk.String(), nodePoolDiskSize.String())
-			// To update the PVCs we need to temporarily delete the StatefulSet while allowing the pods to continue to run
-			r.logger.Info("Deleting statefulset while orphaning pods " + existing.Name)
-			opts := client.DeleteOptions{}
-			client.PropagationPolicy(metav1.DeletePropagationOrphan).ApplyToDelete(&opts)
-			if err := r.Delete(r.ctx, existing, &opts); err != nil {
-				r.logger.Info("Failed to delete statefulset" + existing.Name)
-				return result, err
-			}
-
-			// Identify the PVC for each statefulset pod and patch with the new size
-			for i := 0; i < int(*existing.Spec.Replicas); i++ {
-				clusterName := r.instance.Name
-				claimName := fmt.Sprintf("data-%s-%s-%d", clusterName, nodePool.Component, i)
-				var pvc corev1.PersistentVolumeClaim
-				nsn := types.NamespacedName{
-					Namespace: existing.Namespace,
-					Name:      claimName,
-				}
-				if err := r.Get(r.ctx, nsn, &pvc); err != nil {
-					r.logger.Info("Failed to get pvc" + pvc.Name)
-					return result, err
-				}
-
-				pvc.Spec.Resources.Requests["storage"] = nodePoolDiskSize
-
-				if err := r.Update(r.ctx, &pvc); err != nil {
-					r.logger.Error(err, fmt.Sprintf("Failed to resize statefulset pvc %s", pvc.Name))
-					r.recorder.AnnotatedEventf(r.instance, annotations, "Warning", "PVC", "Failed to Resize %s/%s", existing.Namespace, existing.Name)
-					return result, err
-				}
-			}
-			// STS will be recreated by the normal reconcile below
 		}
 	}
 
@@ -511,4 +466,73 @@ func (r *ClusterReconciler) handlePDB(nodePool *opsterv1.NodePool) (*ctrl.Result
 		}
 		return r.ReconcileResource(&pdb, reconciler.StateAbsent)
 	}
+}
+
+func (r *ClusterReconciler) maybeUpdateVolumes(existing *appsv1.StatefulSet, nodePool opsterv1.NodePool) error {
+	if nodePool.DiskSize == "" { // Default case
+		nodePool.DiskSize = builders.DefaultDiskSize
+	}
+
+	// If we are changing from ephemeral storage to persistent
+	// just delete the statefulset and let it be recreated
+	if len(existing.Spec.VolumeClaimTemplates) < 1 {
+		if err := r.deleteSTSWithOrphan(existing); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	existingDisk := lo.FromPtr(existing.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage())
+	nodePoolDiskSize, err := resource.ParseQuantity(nodePool.DiskSize)
+	if err != nil {
+		r.logger.Error(err, fmt.Sprintf("Invalid diskSize '%s' for nodepool %s", nodePool.DiskSize, nodePool.Component))
+		return err
+	}
+
+	if existingDisk.Equal(nodePoolDiskSize) {
+		r.logger.Info("The existing disk size " + existingDisk.String() + " is same as passed in disk size " + nodePoolDiskSize.String())
+		return nil
+	}
+	r.logger.Info("Disk sizes differ for nodePool %s: current: %s, desired: %s", nodePool.Component, existingDisk.String(), nodePoolDiskSize.String())
+	annotations := map[string]string{"cluster-name": r.instance.GetName()}
+	r.recorder.AnnotatedEventf(r.instance, annotations, "Normal", "PVC", "Starting to resize PVC %s/%s from %s to  %s ", existing.Namespace, existing.Name, existingDisk.String(), nodePoolDiskSize.String())
+	// To update the PVCs we need to temporarily delete the StatefulSet while allowing the pods to continue to run
+	if err := r.deleteSTSWithOrphan(existing); err != nil {
+		return err
+	}
+
+	// Identify the PVC for each statefulset pod and patch with the new size
+	for i := 0; i < int(lo.FromPtrOr(existing.Spec.Replicas, 1)); i++ {
+		clusterName := r.instance.Name
+		claimName := fmt.Sprintf("data-%s-%s-%d", clusterName, nodePool.Component, i)
+		var pvc corev1.PersistentVolumeClaim
+		nsn := types.NamespacedName{
+			Namespace: existing.Namespace,
+			Name:      claimName,
+		}
+		if err := r.Get(r.ctx, nsn, &pvc); err != nil {
+			r.logger.Info("Failed to get pvc" + pvc.Name)
+			return err
+		}
+
+		pvc.Spec.Resources.Requests["storage"] = nodePoolDiskSize
+
+		if err := r.Update(r.ctx, &pvc); err != nil {
+			r.logger.Error(err, fmt.Sprintf("Failed to resize statefulset pvc %s", pvc.Name))
+			r.recorder.AnnotatedEventf(r.instance, annotations, "Warning", "PVC", "Failed to Resize %s/%s", existing.Namespace, existing.Name)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ClusterReconciler) deleteSTSWithOrphan(existing *appsv1.StatefulSet) error {
+	r.logger.Info("Deleting statefulset while orphaning pods " + existing.Name)
+	opts := client.DeleteOptions{}
+	client.PropagationPolicy(metav1.DeletePropagationOrphan).ApplyToDelete(&opts)
+	if err := r.Delete(r.ctx, existing, &opts); err != nil {
+		r.logger.Info("Failed to delete statefulset" + existing.Name)
+		return err
+	}
+	return nil
 }

